### ISSUE TITLE
App data was all deleted on restart of Umbrel

### ISSUE THREAD

**Issue by maltokyo** [getumbrel/umbrel#408]
This is somewhat related to: https://github.com/getumbrel/umbrel/issues/382

I installed BTCPay Server app, added some wallets there to check balance etc.

I then ran ./scripts/stop as I needed to reboot the server.

After reboot, and running ./scripts/start again, I was surprised that docker ps showed my umbrel "apps" restarting strangely like this:

```
CONTAINER ID   IMAGE                               COMMAND                  CREATED          STATUS                            PORTS                                                                                           NAMES
0c0c0a281aee   lncm/specter-desktop:v0.10.4        "/usr/local/bin/pyth…"   12 minutes ago   Restarting (1) 43 seconds ago                                                                                                     specter-desktop_web_1
e624620b292e   btcpayserver/btcpayserver:1.0.6.4   "dotnet BTCPayServer…"   12 minutes ago   Restarting (139) 56 seconds ago                                                                                                   btcpay-server_web_1
fcf3ecfbb2f9   postgres:9.6.20                     "docker-entrypoint.s…"   12 minutes ago   Restarting (1) 16 seconds ago                                                                                                     btcpay-server_postgres_1
d104dbe52535   nicolasdorier/nbxplorer:2.1.46      "dotnet NBXplorer.dll"   12 minutes ago   Restarting (139) 56 seconds ago                                                                                                   btcpay-server_nbxplorer_1
fbe28dfb9ed2   getumbrel/btc-rpc-explorer:v2.1.0   "docker-entrypoint.s…"   12 minutes ago   Up 12 minutes                     0.0.0.0:3002->3002/tcp                                                                          btc-rpc-explorer_web_1
b68d67144ba0   getumbrel/electrs:v0.8.6            "electrs"                12 minutes ago   Up 12 minutes                     4224/tcp, 0.0.0.0:50001->50001/tcp                                                              electrs
79b28cc755d3   lncm/lnd:v0.11.1                    "/usr/local/bin/lnd"     12 minutes ago   Up 12 minutes                     0.0.0.0:8080->8080/tcp, 0.0.0.0:9735->9735/tcp, 0.0.0.0:10009->10009/tcp, 9911/tcp              lnd
4f8b088b1f03   nginx:1.17.8                        "nginx -g 'daemon of…"   12 minutes ago   Up 12 minutes                     80/tcp                                                                                          nginx
3d0ec269a8c1   getumbrel/middleware:v0.1.7         "docker-entrypoint.s…"   12 minutes ago   Up 12 minutes                     3006/tcp                                                                                        middleware
a4d85ee4f0d2   lncm/bitcoind:v0.20.1               "bitcoind -zmqpubraw…"   12 minutes ago   Up 12 minutes                     8080/tcp, 8332/tcp, 18332-18333/tcp, 18443-18444/tcp, 28332-28333/tcp, 0.0.0.0:8333->8333/tcp   bitcoin
761ed9dabdb0   getumbrel/manager:v0.2.9            "docker-entrypoint.s…"   12 minutes ago   Up 12 minutes                     3006/tcp                                                                                        manager
ae4dd13682db   getumbrel/tor:v0.4.1.9              "tor"                    12 minutes ago   Up 12 minutes                     9051/tcp, 127.0.0.1:9050->9050/tcp, 29050-29051/tcp                                             tor
224d946187f9   getumbrel/dashboard:v0.3.14         "docker-entrypoint.s…"   12 minutes ago   Up 12 minutes                     3004/tcp                                                                                        dashboard
e506d16366ec   getumbrel/frontail:v4.9.1           "/frontail/docker-en…"   12 minutes ago   Up 12 minutes                     9001/tcp                                                                                        frontail
```

So I went into the logs of BTCPay to have a look and found this:

```
Unhandled exception. System.TypeInitializationException: The type initializer for 'BTCPayServer.BTCPayDefaultSettings' threw an exception.
 ---> System.UnauthorizedAccessException: Access to the path '/data/.btcpayserver' is denied.
 ---> System.IO.IOException: Permission denied
   --- End of inner exception stack trace ---
   at System.IO.FileSystem.CreateDirectory(String fullPath)
   at System.IO.Directory.CreateDirectory(String path)
   at StandardConfiguration.DefaultDataDirectory.GetDirectory(String appDirectory, String subDirectory, Boolean createIfNotExists)
   at BTCPayServer.BTCPayDefaultSettings..cctor() in /source/BTCPayServer.Common/BTCPayNetwork.cs:line 26
   --- End of inner exception stack trace ---
   at BTCPayServer.BTCPayDefaultSettings.GetDefaultSettings(NetworkType chainType) in /source/BTCPayServer.Common/BTCPayNetwork.cs:line 40
   at BTCPayServer.BTCPayNetworkProvider.InitBitcoin() in /source/BTCPayServer.Common/BTCPayNetworkProvider.Bitcoin.cs:line 12
   at BTCPayServer.BTCPayNetworkProvider..ctor(NetworkType networkType) in /source/BTCPayServer.Common/BTCPayNetworkProvider.cs:line 85
   at BTCPayServer.Configuration.DefaultConfiguration.CreateCommandLineApplicationCore() in /source/BTCPayServer/Configuration/DefaultConfiguration.cs:line 16
   at StandardConfiguration.DefaultConfiguration.CreateCommandLineApplication()
   at StandardConfiguration.DefaultConfiguration.CreateConfiguration(String[] args)
   at BTCPayServer.Program.Main(String[] args) in /source/BTCPayServer/Program.cs:line 28
```

Basically, the cause was that the entire content of the app directory has been deleted...! I did not uninstall the app in the middle, only a server restart.  Seems like a critical bug.

I would propose that app data is never touched, even if I do choose to uninstall an app, or at least give the option to keep the data.  (Again, in the case above, I did not even uninstall)

**Comment by maltokyo** [comment]
@mayankchhabra @lukechilds - would you please help, with any advice?  I don't know what to do, I cant trust the setup of BTCPay, etc will survive, so not trying it anymore.

**Comment by lukechilds** [comment]
Thanks for reporting @maltokyo! Sorry for the delay, this slipped through the cracks.

You're running a custom install, right? Not Umbrel OS?

When you say

>the cause was that the entire content of the app directory has been deleted...!

exactly which directory was missing?
- `app-data/`
- `app-data/btcpay-server/`
- a deeper level?

**Comment by maltokyo** [comment]
Hi @lukechilds 
The app-data for was empty. 

Correct, I'm running on a fresh install of Debian 10.  Only docker and Umbrel installed, no other services working on this PC.

**Comment by lukechilds** [comment]
Ok, thanks!

>I then ran ./scripts/stop as I needed to reboot the server.

Why did you need to reboot, had you just done some maintenance? dist upgrade? Also had you done an Umbrel update via CLI?

**Comment by maltokyo** [comment]
I needed to move it to another room that's all ;) no other updates etc.

**Comment by maltokyo** [comment]
And previously I used to update always via CLI. Until you told me in another issue here that it's bad practice. So for last two versions, I've used only web UI to upgrade. Thank you.

**Comment by lukechilds** [comment]
Have you updated since you ran into the issue?

Do you know exactly what Umbrel version you were on when the issue occurred? 

Also, do you know if you've restarted before without experiencing the issue? Or this is the first time you rebooted so you can't tell?

**Comment by maltokyo** [comment]
It was the first time to restart after app store was introduced. But restarts before then had no issues at all. 

it happened on the last version. 0.3.1 (?) (when the apps were just released, I was trying it out the same day it was released I think, and this was a couple of days after that).

**Comment by lukechilds** [comment]
Ok thanks, I don't seem to be able to recreate the issue.

I'm:
- Installing BTCPay
- Stopping Umbrel
- Rebooting
- Starting Umbrel
- Checking app-data/btcpay-server still exists

If you have the time would you be willing to quickly try installing the app again and rebooting and seeing if the issue reoccurs?

Also do you have any custom bind mounts setup to mount app-data on a different block device or something?

**Comment by maltokyo** [comment]
Hi Luke, Actually on the new version I tried to recreate this again on the new version. But seems working now. No custom bind mounts or anything fancy.
Shall we close for now? Nobody else seemed to have the issue...
Thanks for checking though.

**Comment by lukechilds** [comment]
Ok closing this for now, feel free to re-open if you the issue reoccurs.
